src/cryptonote_core/service_node_list.cpp

```diff
@@ -312,7 +320,114 @@ namespace service_nodes
 
 
     m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
+    uint64_t swarm_id = iter->second.swarm_id;
     m_service_nodes_infos.erase(iter);
+    check_and_process_swarm_decomission(swarm_id);
+  }
+
+  void service_node_list::check_and_process_swarm_decomission(uint64_t swarm_id)
+  {
+    size_t count = 0;
+    for (const auto& entry : m_service_nodes_infos)
+      if (entry.second.swarm_id == swarm_id)
+        count++;
+    if (count == MIN_SWARM_SIZE - 1)
+    {
+      // XXX XXX XXX XXX
+      // All nodes should fetch data from this swarm at this point only.
+      // The registered nodes will eventually disappear and after this point,
+      // it is already considered gone. It only exists to retrieve data from.
+      //
+      // XXX XXX XXX XXX
+      //         internship optimization hardfork idea:
+      //         when nodes have been decomissioned for more than 10 blocks,
+      //         move the nodes into new swarms immediately.
+    }
+  }
+
+  uint64_t service_node_list::get_new_node_swarm_id(uint64_t block_height, uint32_t index) const
+  {
+    // XXX XXX XXX improvement proposal: have the swarm id be "remembered" from
+    // the last time, if the new node was already registered before. This would
+    // have huge performance gains preventing nodes from having to resync every
+    // time they rejoin the network.
```

Comment:
I think this case would be best handled by adding that grace period after deregistration/expiration, so they're still receiving and pushing data on the network. Otherwise might be a slight problem in having to deal with stale data they're holding in the time they take to re-register (in the case of deregistration you can imagine this could be longer, if say they decide to transfer funds back in to re-stake).

Reply:
Agreed that having a way to restake and prevent expiration from happening at all would be ideal.

I'll have a think about some ways to do this.

Reply:
Ideal would actually be exposing the key image in the staking transaction, then as long as that key image doesn't appear in any transaction, the node doesn't expire. The moment the key image appears in a TX, the node expires. @KeeJef this may interest you, not sure if there are resources available for this to go in the next hard fork

Reply:
I think this might be a lot easier than it seems. On service node list side, the criteria for expired nodes is 30 days + grace. When processing registration, check the service node in our list and if the registration is within that grace period, accept it.

Then you just change the sanity checks for submitting registration to check that the registration height is in that grace period.

Reply:
nice one